docs/openapi-docs/src/main/scala/tapir/docs/openapi/ObjectSchemasForEndpoints.scala

```diff
@@ -1,76 +1,91 @@
 package tapir.docs.openapi
 
-import tapir.Schema.SObjectInfo
+import tapir.Schema.{SObject, SObjectInfo}
+import tapir.openapi.OpenAPI.ReferenceOr
 import tapir.openapi.{Schema => OSchema}
 import tapir.{Schema => SSchema, _}
 
+import scala.collection.immutable
+
 object ObjectSchemasForEndpoints {
   type SchemaKey = String
-  type SchemaKeys = Map[SObjectInfo, (SchemaKey, OSchema)]
+  type SchemaKeys = Map[SObjectInfo, (SchemaKey, ReferenceOr[OSchema])]
 
   def apply(es: Iterable[Endpoint[_, _, _]]): SchemaKeys = {
-    val infosToSchema = foldMaps(es.map(e => forInput(e.input) ++ forIO(e.errorOutput) ++ forIO(e.output)))
-
-    // avoiding name clashes if two schemas
-    case class SchemaKeyAssignment(keyToInfo: Map[SchemaKey, SObjectInfo], schemas: SchemaKeys)
+    val schemas = foldLists(es.map(e => forInput(e.input) ++ forIO(e.errorOutput) ++ forIO(e.output)))
+    val infos = schemas.collect { case SObject(info, _, _) => info }
+    val infoToKey = calculateUniqueKeys(infos)
+    val sschemaToOSchema = new SSchemaToOSchema(infoToKey.map { case (k, v) => k.fullName -> v })
+    val infosToSchema = schemas.flatMap(objectSchemaToOSchema(sschemaToOSchema)).toMap
+    infosToSchema.map { case (k, v) => k -> (infoToKey(k) -> v) }
+  }
 
-    val result = infosToSchema.foldLeft(SchemaKeyAssignment(Map.empty, Map.empty)) {
-      case (SchemaKeyAssignment(keyToInfo, schemas), (objectInfo, schema)) =>
-        var key = objectInfo.shortName
-        var i = 0
-        while (keyToInfo.contains(key) && !keyToInfo.get(key).contains(objectInfo)) {
-          i += 1
-          key = objectInfo.shortName + i
-        }
+  private def calculateUniqueKeys(infos: immutable.Seq[SObjectInfo]) = {
+    case class SchemaKeyAssignment1(keyToInfo: Map[SchemaKey, SObjectInfo], infoToKey: Map[SObjectInfo, SchemaKey])
+    infos
+      .foldLeft(SchemaKeyAssignment1(Map.empty, Map.empty)) {
+        case (SchemaKeyAssignment1(keyToInfo, infoToKey), objectInfo) =>
+          var key = objectInfo.shortName
+          var i = 0
+          while (keyToInfo.contains(key) && !keyToInfo.get(key).contains(objectInfo)) {
+            i += 1
+            key = objectInfo.shortName + i
+          }
 
-        SchemaKeyAssignment(
-          keyToInfo + (key -> objectInfo),
-          schemas + (objectInfo -> ((key, schema)))
-        )
+          SchemaKeyAssignment1(
+            keyToInfo + (key -> objectInfo),
+            infoToKey + (objectInfo -> key)
+          )
+      }
+      .infoToKey
+  }
+  private def objectSchemaToOSchema(sschemaToOSSchema: SSchemaToOSchema)(schema: SSchema): Map[SObjectInfo, ReferenceOr[OSchema]] = {
+    schema match {
+      case SSchema.SObject(info, _, _) =>
+        Map(info -> sschemaToOSSchema(schema))
+      case _ => Map.empty
     }
+  }
 
-    result.schemas
+  private def filterIsObjectSchema(schema: SSchema): List[SSchema] = {
```

Comment:
you can probably just return a `List[SSchema.SObject]` and later avoid `.collect`

Reply:
I thought about it but wasn't sure about that one since it is still hard for me to distinguish sometimes between the code which is written for the future usage and the code which can be just simplified